webpack/components/Search/Search.js

```diff
@@ -0,0 +1,131 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import React, { Component } from 'react';
+import { ControlLabel } from 'react-bootstrap';
+import PropTypes from 'prop-types';
+import { translate as __ } from 'foremanReact/common/I18n';
+import TypeAhead from '../TypeAhead';
+import api from '../../services/api';
+import { stringIncludes } from './helpers';
+import {
+  AUTOSEARCH_DELAY,
+  AUTOSEARCH_WHILE_TYPING,
+} from '../../scenes/Settings/SettingsConstants';
+
+class Search extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { items: [], typingTimeout: 0, autoSearchOnEmpty: false };
+    this.onInputUpdate = this.onInputUpdate.bind(this);
+    this.onSearch = this.onSearch.bind(this);
+  }
+
+  componentDidMount() {
+    this.onInputUpdate();
+    this.props.loadSetting(AUTOSEARCH_DELAY);
+    this.props.loadSetting(AUTOSEARCH_WHILE_TYPING);
+  }
+
+  async onInputUpdate(searchTerm = '') {
+    const { getAutoCompleteParams, settings: { autoSearchEnabled }, patternfly4 } = this.props;
+    const items = this.state.items.filter(({ text }) => stringIncludes(text, searchTerm));
+
+    if (items.length !== this.state.items.length) {
+      this.setState({ items });
+    }
+
+    const params = getAutoCompleteParams(searchTerm);
+    const autoCompleteParams = [
+      params.endpoint,
+      params.headers || {},
+      params.params || {},
+    ];
+
+    if (autoCompleteParams[0] !== '') {
+      const { data } = await api.get(...autoCompleteParams);
+      this.setState({
+        items: data.filter(({ error }) => !error).map(({ label }) => ({
+          text: label.trim(),
+        })),
+      });
+    }
+
+    if (autoSearchEnabled &&
+        patternfly4 &&
+        (searchTerm.length > 0 || this.state.autoSearchOnEmpty)) {
+      this.autoSearch(searchTerm);
+    }
+
+    // Prevent empty search when page first loads, but search when input is cleared after that
+    this.setState({ autoSearchOnEmpty: true });
+  }
+
+  onSearch(search) {
+    if (this.props.updateSearchQuery) this.props.updateSearchQuery(search);
```

Comment:
This works, but we will need to configure eslint to play nice with this syntax:
```
/home/vagrant/katello/webpack/components/Search/Search.js
   62:5  error  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  112:3  error  'updateSearchQuery' PropType is defined but prop is never used         react/no-unused-prop-types
```

How about I just keep the regular syntax for now until we figure out how to support optional chaning with methods for eslint? I don't think it's worth the effort at this time.

Reply:
@johnpmitsch looks like we may have to [add](https://stackoverflow.com/questions/57378411/eslint-rule-for-optional-chaining-nullish-coalescing-operator) the Babel plugin `babel-eslint`.  I'm fine with leaving as-is for now.

Reply:
In general I think you can still do `?.` without eslint changes, it was just confused by `?.(functionCall)`

Reply:
@jeremylenz do you have a suggested change in mind? I get its improved syntax, but considering it will do the same thing at runtime and there isn't a lot of excess code here, I don't really want to spend a lot of time trying to update eslint or working around the existing rules here. If you want to suggest something that passes eslint and tests as-is, I can update to that.

Reply:
@johnpmitsch sorry, no I didn't mean make any changes to this!  I just meant that in general I wouldn't want you to hesitate to use optional chaining elsewhere, as long as there are no function calls that trip up eslint.

Reply:
> @johnpmitsch sorry, no I didn't mean make any changes to this! I just meant that in general I wouldn't want you to hesitate to use optional chaining elsewhere, as long as there are no function calls with arguments that trip up eslint.
> 
> ```js
> this?.props?.handleChange() // ok
> ```
> 
> ```js
> this?.props?.handleChange?.(arg) // confuses eslint
> ```

ah sorry I misunderstood, I concluded the same and plan to use when there are no function calls as well! :+1: